storage/fanout.go

```diff
@@ -569,11 +593,14 @@ func (h *genericSeriesSetHeap) Pop() interface{} {
 	return x
 }
 
-// ChainedSeriesMerge returns single series from many same series by chaining samples together.
+// ChainingSeriesMerge returns single series from many same series by chaining samples together.
 // In case of the timestamp overlap, the first overlapped sample is kept and the rest samples with the same timestamps
 // are dropped. We expect the same labels for each given series.
-// TODO(bwplotka): This has the same logic as tsdb.verticalChainedSeries. Remove this in favor of ChainedSeriesMerge in next PRs.
-func ChainedSeriesMerge(s ...Series) Series {
+//
+// This works the best with replicated series, where data from two series are exactly the same. This does not work well
+// with "almost" the same data, e.g. from 2 Prometheus HA replicas. This is fine, since from the Prometheus perspective
+// this never happens.
+func ChainingSeriesMerge(s ...Series) Series {
```

Comment:
While the name was changed back in #7005, I see that this also merges the overlapping data. Chaining sounds more like non-overlapping and attaching results directly.

Reply:
Cool, so what would you suggest? Because to me this algo is exactly doing that. It does not merge anything.. It's just appends ALL together not caring about any scrape interval at all....

Reply:
https://github.com/prometheus/prometheus/blob/a2000f39a70983ec1f25fa0c6bc43e68a6384d1f/storage/fanout.go#L681-L685

This tells me that we are actually vertically merging it here by dropping the samples with same timestamp.